prometheus/counter.go

```diff
@@ -40,6 +41,16 @@ type Counter interface {
 	// Add adds the given value to the counter. It panics if the value is <
 	// 0.
 	Add(float64)
+	// AddWithExemplar works like Add but also replaces the currently saved
+	// exemplar (possibly none) with a new one, created from the provided
+	// value, the current time as timestamp, and the provided labels. Empty
+	// Labels will lead to a valid (label-less) exemplar. But if Labels is
+	// nil, the current exemplar is left in place. This method panics if the
+	// value is < 0, if any of the provided labels are invalid, or if the
+	// provided labels contain more than 64 runes in total. AddWithExemplar
```

Comment:
Mightn't hurt to put numbers on this. What you and I consider costly for instrumentation isn't the same as everyone else, given that e.g. sending a udp packet on every event or using client-side quantiles is common.

Reply:
Hmm, those numbers would be quite involved. It's not just runtime, but also that the creation of the Exemplar needs a number of allocations.

Re-reading the current wording of the doc comment, I find it now a bit confusing myself. It would probably best to explain the usual strategy for really hot paths (stochastically do some observations with and most without exemplar or even do some more sophisticated adaptive rate limiting), but then it would get way too involved for a doc comment.

I removed the sentence about the performance impact for now. This is after all still an experiment. (I'll mark it as such in the changelog.)

Reply:
My thought was mainly to avoid scaring people away from using this for reasons that they don't actually care about. If they really care about a hot path, they'll probably have their own micro-benchmarks anyway.